src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+            sourceFile,
+            context.program.getTypeChecker());
+
+        context.cancellationToken.throwIfCancellationRequested();
+        return scopes.map((scope, i) => {
+            const errors = errorsPerScope[i];
+            if (errors.length) {
+                return {
+                    scope,
+                    scopeDescription: getDescriptionForScope(scope),
+                    errors
+                };
+            }
+            return extractFunctionInScope(target, scope, usagesPerScope[i], targetRange, context);
+        });
+    }
+
+    function getDescriptionForScope(s: Scope) {
+        if (isFunctionLike(s)) {
+            switch (s.kind) {
+                case SyntaxKind.Constructor:
+                    return "constructor";
+                case SyntaxKind.FunctionExpression:
+                    return s.name
+                        ? `function expression ${s.name.getText()}`
+                        : "anonymous function expression";
+                case SyntaxKind.FunctionDeclaration:
+                    return `function ${s.name.getText()}`;
+                case SyntaxKind.ArrowFunction:
+                    return "arrow function";
+                case SyntaxKind.MethodDeclaration:
+                    return `method ${s.name.getText()}`;
+                case SyntaxKind.GetAccessor:
+                    return `get ${s.name.getText()}`;
+                case SyntaxKind.SetAccessor:
+                    return `set ${s.name.getText()}`;
+            }
+        }
+        else if (isModuleBlock(s)) {
+            return `namespace ${s.parent.name.getText()}`;
+        }
+        else if (isClassLike(s)) {
+            return s.kind === SyntaxKind.ClassDeclaration
+                ? `class ${s.name.text}`
+                : s.name.text
+                    ? `class expression ${s.name.text}`
+                    : "anonymous class expression";
+        }
+        else if (isSourceFile(s)) {
+            return `file '${s.fileName}'`;
+        }
+        else {
+            return "unknown";
+        }
+    }
+
+    export function extractFunctionInScope(
+        node: Node,
+        scope: Scope,
+        { usages: usagesInScope, substitutions }: ScopeUsages,
+        range: TargetRange,
+        context: RefactorContext): ExtractResultForScope {
+
+        const checker = context.program.getTypeChecker();
+
+        const changeTracker = textChanges.ChangeTracker.fromRefactorContext(context);
+
+        // Make a unique name for the extracted function
+        let functionNameText = "newFunction";
+        if (scope.locals && scope.locals.has(functionNameText)) {
+            let i = 1;
+            while (scope.locals.has(functionNameText = `newFunction_${i}`)) {
+                i++;
+            }
+        }
+
+        const functionName = createIdentifier(functionNameText);
+        // Currently doesn't get populated, but we might try to infer from this at some point

```

Comment:
Does this comment mean that we don't support extracting portions of generic methods that depend on type parameters?